sky/backends/cloud_vm_ray_backend.py

```diff
@@ -2790,8 +2780,30 @@ def _exec_code_on_head(
             mkdir_code = (f'{cd} && mkdir -p {remote_log_dir} && '
                           f'touch {remote_log_path}')
             code = job_lib.JobLibCodeGen.queue_job(job_id, job_submit_cmd)
+
             job_submit_cmd = mkdir_code + ' && ' + code
 
+            if spot_task is not None:
+                # Add the spot job to spot queue table.
+                resources_str = backend_utils.get_task_resources_str(spot_task)
+                spot_codegen = spot_lib.SpotCodeGen()
+                spot_name = spot_task.name
+                assert spot_name is not None, spot_task
+                spot_code = spot_codegen.set_pending(job_id, spot_name,
+                                                     resources_str)
+                # Set the spot job to PENDING state to make sure that this spot
+                # job appears in the `sky spot queue`, when there are already 16
+                # controller process jobs running on the controller VM with 8
+                # CPU cores.
+                # The spot job should be set to PENDING state after the
+                # controller process job has been queued, as our skylet on spot
```

Comment:
How does our own job queue work? Does it automatically detect there are too many `ray job submit`? Or does it rely on some other API calls / tables?

Reply:
It relies on another `pending` table. The `ray job submit` command will be stored in the `pending` table. When the latest job get the required resources, our scheduler will run the `ray job submit` command for the first job in the pending table, i.e. the `ray job submit` will be delayed until the prior job is scheduled.